content/concepts/peer-routing.md

```diff
@@ -3,6 +3,82 @@ title: Peer Routing
+
+The Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. This guarantees that any node can be located by it's peer ID.
+
+Let's see how Bob can locate Alice in this example. Bob knows Alices peerID, but he needs to get the value - Alices IP address. From subtree with prefix `1` Bob knows peer with prefix `100`. He sends a message to their address, where he asks for Alices address.
+
+![1-hop](1-hop.png)
+
+Peer `100` doesn't have Alice's key. But they have key of node `110` from their subtree which locates Alice better! And we are doing second hop:
+
+![2-hop](2-hop.png)
+
+Peer `110` also doesn't have Alice's key, but they again locate Alice better and send request to node with prefix `1111`.
+
+![3-hop](3-hop.png)
+
+Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:
+
+![final-hop](final-hop.png)
+
+Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.
+
+### Routing table
+
+The place where one node stores info about other nodes called routing table. The routing table is a binary tree whose leaves are `k-buckets`. Let's boil it down.
+
+What are `k-buckets`? Remember our initial tree? Here it is. ![Kademlia-system](Kademlia-system.png) See this circled sub-trees? Also, remember that Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
+
+![k-buckets](k-buckets.png)
+
+Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.
+
+## Dynamic DHT
+
+Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
+
+### Leaving
+
+Let us assume that we have some known parameter `k` that represents the number of computers such that it is extremely unlikely that all of those computers will leave the network in the same hour. For example, the authors of the original Kademlia paper experimentally determined that their `k` should be 20.
+
+Next, our `k-bucket` should, instead of storing just 1 computer, store `k` computers within that `k-bucket’s` range. Thus, with high likelihood, there will always be at least one computer online in each of a computer’s `k-buckets`.
+
+And, when hopping, instead of asking one computer from `k-bucket` we will be asking `k`. Some may not reply, but not all of them.
```

Comment:
```suggestion
And, when hopping, instead of asking one computer from the `k-bucket` we will be asking `k` number of peers. Some may not reply, but not all.
```